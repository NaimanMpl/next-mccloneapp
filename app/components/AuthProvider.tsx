'use client';
import { createContext, Dispatch, ReactNode, SetStateAction, useContext, useEffect, useState } from "react";
import { UserPayload } from "../models/user.model";
import { getCurrentUser } from "../services/authservice";

interface IAuthContext {
  user: UserPayload | null,
  setUser: Dispatch<SetStateAction<UserPayload | null>>
}

export const AuthContext = createContext<IAuthContext | undefined>(undefined);

interface Props {
  children: ReactNode
}

export const useAuth = () => { 
  const context = useContext(AuthContext);
  if (context === undefined) {
    throw new Error('useAuth must be used within a AuthProvider');
  } 
  return context;
}

export const AuthProvider = ({ children }: Props) => {
  
  const [ user, setUser ] = useState<UserPayload | null>(null);

  useEffect(() => {
    (async () => {
      const currentUser = await getCurrentUser();
      if (currentUser == null) return;

      setUser({ email: currentUser.email, name: currentUser.name, id: currentUser.id, skin: currentUser.skin, role: currentUser.role, admin: currentUser.admin });
    })();
  }, []);

  return (
    <AuthContext.Provider value={{ user, setUser }}>
      {children}
    </AuthContext.Provider>
  )

}
