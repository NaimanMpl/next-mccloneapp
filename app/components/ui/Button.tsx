import React from 'react';

interface ButtonProps {
  label: string,
  backgroundColor: string,
  color: string,
  fullWidth?: boolean,
  loading?: boolean,
  submit?: boolean
}

const Button = ({ backgroundColor, color, loading = false, label, fullWidth = true, submit = true }: ButtonProps) => {
  const textColor = `text-${color}`;
  return (
    <button
      type={submit ? 'submit' : undefined}
      className={`mt-4 ${fullWidth ? 'w-full' : 'px-10'} ${backgroundColor} rounded-md py-2 font-semibold ${textColor}`}
      disabled={loading}
    >
      {label}
    </button>
  )
}

export default Button;