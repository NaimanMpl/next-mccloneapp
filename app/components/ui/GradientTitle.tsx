import { cn } from '@/lib/utils'
import { ReactNode } from 'react'

interface GradientTitleProps {
  children: ReactNode,
  className?: string
}

const GradientTitle = ({ children, className }: GradientTitleProps) => {
  return (
    <h2 
      className={cn(`pointer-events-none z-10 whitespace-pre-wrap bg-gradient-to-r from-[#000000] from-65% to-neutral-200/60 bg-clip-text text-center text-7xl font-semibold leading-none text-transparent dark:from-white dark:to-slate-900/10 mobile:text-5xl mobile:from-50%`, className)}
    >
      {children}
    </h2>
  )
}

export default GradientTitle