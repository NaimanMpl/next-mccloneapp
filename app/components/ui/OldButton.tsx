import React from 'react';

interface OldButtonProps {
  label: string,
  backgroundColor: string,
  color: string,
  fullWidth?: boolean,
  loading?: boolean,
  submit?: boolean
}

const OldButton = ({ backgroundColor, color, loading = false, label, fullWidth = true, submit = true }: OldButtonProps) => {
  const textColor = `text-${color}`;
  return (
    <button
      type={submit ? 'submit' : undefined}
      className={`${fullWidth ? 'w-full' : 'px-10'} ${backgroundColor} rounded-md py-2 font-semibold ${textColor}`}
      disabled={loading}
    >
      {label}
    </button>
  )
}

export default OldButton;