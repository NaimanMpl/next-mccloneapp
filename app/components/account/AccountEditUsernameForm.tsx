import { Button } from '@/components/ui/button';
import { Card, CardContent, CardDescription, CardFooter, CardHeader, CardTitle } from '@/components/ui/card';
import { Input } from '@/components/ui/input';
import React from 'react';

interface AccountEditUsernameFormProps {
  username: string
}

const AccountEditUsernameForm = ({ username }: AccountEditUsernameFormProps) => {
  return (
    <Card>
      <CardHeader>
        <CardTitle className='text-xl font-semibold'>Nom d'utilisateur</CardTitle>
        <CardDescription>Le nom d'utilisateur est utilisé pour vous différencier des autres joueurs.</CardDescription>
      </CardHeader>
      <CardContent>
        <form>
          <Input className='mt-3 w-80' type='text' defaultValue={username} />
        </form>
      </CardContent>
      <CardFooter className='border-t border-border px-6 py-2'>
        <div className='flex justify-between items-center w-full'>
          <CardDescription>Veuillez ne pas dépasser 20 caractères.</CardDescription>
          <Button>Sauvegarder</Button>
        </div>
      </CardFooter>
    </Card>
  )
}

export default AccountEditUsernameForm