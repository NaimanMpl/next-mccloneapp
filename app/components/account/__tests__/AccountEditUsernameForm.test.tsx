import mockAxios from "@/__mocks__/axios";
import { Toaster } from "@/components/ui/toaster";
import { render, screen, waitFor } from "@testing-library/react";
import user from '@testing-library/user-event';
import { AuthProvider } from "../../AuthProvider";
import AccountEditUsernameForm from "../AccountEditUsernameForm";

jest.mock('axios');

describe('<AccountEditUsernameFOrm>', () => { 

  afterEach(() => {
    mockAxios.reset();
    user.setup();
  })

  it('should render successfully', () => {

    const username = "John";

    const form = render(
      <AuthProvider>
        <AccountEditUsernameForm username={username} />
      </AuthProvider>
    );

    const [ usernameInput ] = screen.getAllByRole('textbox');

    expect(usernameInput).toHaveValue(username);
    expect(form).toMatchSnapshot();
  });

  it('should change input value on successfull rename api call', async () => {

    const username = "John";
    const newUsername = "John Doe";
    mockAxios.patch.mockResolvedValue({ username: newUsername });

    render(
      <AuthProvider>
        <AccountEditUsernameForm username={username} />
      </AuthProvider>
    );

    const [ usernameInput ] = screen.getAllByRole('textbox');

    await user.clear(usernameInput);
    await user.type(usernameInput, newUsername);

    const submitButton = screen.getByText('Sauvegarder');
    await user.click(submitButton);

    await waitFor(() => {
      expect(mockAxios.patch).toHaveBeenCalledWith(`/api/auth/me?username=${newUsername}`);
      expect(usernameInput).toHaveValue(newUsername);
    });
    
  });

  it('should throw a warning toast if the user keeps his name unchanged', async () => {

    const username = 'John';
    mockAxios.get.mockResolvedValueOnce({ name: username });

    render(
      <>
        <AuthProvider>
          <AccountEditUsernameForm username={username} />
        </AuthProvider>
        <Toaster />
      </>
    );

    const [ usernameInput ] = screen.getAllByRole('textbox');

    await user.clear(usernameInput);
    await user.type(usernameInput, username);

    const submitButton = screen.getByText('Sauvegarder');
    await user.click(submitButton);

    await waitFor(() => {
      expect(usernameInput).toHaveValue(username);
      // expect(screen.getByText('Pas si vite !')).toBeInTheDocument();
    });
    
  });

});