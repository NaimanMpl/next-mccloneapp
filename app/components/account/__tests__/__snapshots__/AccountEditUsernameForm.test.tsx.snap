// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AccountEditUsernameFOrm> should render successfully 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="rounded-lg border border-border bg-card text-card-foreground shadow-sm"
      >
        <div
          class="flex flex-col space-y-1.5 p-6"
        >
          <h3
            class="tracking-tight text-xl font-semibold"
          >
            Nom d'utilisateur
          </h3>
          <p
            class="text-sm text-muted-foreground"
          >
            Le nom d'utilisateur est utilisé pour vous différencier des autres joueurs.
          </p>
        </div>
        <form>
          <div
            class="p-6 pt-0"
          >
            <div
              class="space-y-2"
            >
              <input
                aria-describedby=":r0:-form-item-description"
                aria-invalid="false"
                class="flex h-10 rounded-md border border-input bg-background px-3 py-2 text-sm ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50 mt-3 w-80"
                id=":r0:-form-item"
                name="username"
                placeholder="Imperator"
                type="text"
                value="John"
              />
            </div>
          </div>
          <div
            class="flex items-center p-6 border-t border-border px-6 py-2"
          >
            <div
              class="flex justify-between items-center w-full"
            >
              <p
                class="text-sm text-muted-foreground"
              >
                Veuillez ne pas dépasser 20 caractères.
              </p>
              <button
                class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-semibold ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 bg-primary text-primary-foreground hover:bg-primary/90 h-10 px-4 py-2"
                type="submit"
              >
                Sauvegarder
              </button>
            </div>
          </div>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="rounded-lg border border-border bg-card text-card-foreground shadow-sm"
    >
      <div
        class="flex flex-col space-y-1.5 p-6"
      >
        <h3
          class="tracking-tight text-xl font-semibold"
        >
          Nom d'utilisateur
        </h3>
        <p
          class="text-sm text-muted-foreground"
        >
          Le nom d'utilisateur est utilisé pour vous différencier des autres joueurs.
        </p>
      </div>
      <form>
        <div
          class="p-6 pt-0"
        >
          <div
            class="space-y-2"
          >
            <input
              aria-describedby=":r0:-form-item-description"
              aria-invalid="false"
              class="flex h-10 rounded-md border border-input bg-background px-3 py-2 text-sm ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50 mt-3 w-80"
              id=":r0:-form-item"
              name="username"
              placeholder="Imperator"
              type="text"
              value="John"
            />
          </div>
        </div>
        <div
          class="flex items-center p-6 border-t border-border px-6 py-2"
        >
          <div
            class="flex justify-between items-center w-full"
          >
            <p
              class="text-sm text-muted-foreground"
            >
              Veuillez ne pas dépasser 20 caractères.
            </p>
            <button
              class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-semibold ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 bg-primary text-primary-foreground hover:bg-primary/90 h-10 px-4 py-2"
              type="submit"
            >
              Sauvegarder
            </button>
          </div>
        </div>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
